File: loc.go
package main

import (
	"bufio"
	"fmt"
	"io"
	"log"
	"net/http"
	"os"
	"os/exec"
	"strconv"
	"strings"
	"time"

	lib "./_lib"
)

var languages []language

func doLocUploadRequest(w http.ResponseWriter, r *http.Request, folder string) {
	folderNum := folder
	folder = "temp/loc/" + folder
	destroyTemporalDir := []string{"rm", "-r", folder}
	createTemporalDir := []string{"mkdir", folder}
	err := lib.GenericCommand(createTemporalDir)
	if err != nil {
		log.Printf("ERROR 1 creating folder %s\n", err)
		e.Error = "Error creating folder " + folder
		lib.SendErrorToClient(w, e)
		return
	}

	// create file
	file, handler, err := r.FormFile("inputFile")
	if err != nil {
		log.Printf("ERROR creating file %s\n", err)
		e.Error = "Error creating file "
		lib.SendErrorToClient(w, e)
		lib.GenericCommand(destroyTemporalDir)
		return
	}
	upload := handler.Filename
	filePath := "./" + folder + "/" + upload
	defer file.Close()
	f, err := os.OpenFile(filePath, os.O_WRONLY|os.O_CREATE, 0666)
	if err != nil {
		log.Printf("ERROR opening uploaded file %s\n", err)
		e.Error = "Error opening " + upload
		lib.SendErrorToClient(w, e)
		lib.GenericCommand(destroyTemporalDir)
		return
	}
	defer f.Close()
	io.Copy(f, file)

	dest := "./" + folder
	//unzipFile := []string{"unzip", filePath, "-d", dest + "/src"}
	unzipFile := []string{"7z", "x", filePath, "-o" + dest + "/src"}
	err = lib.GenericCommand(unzipFile)
	if err != nil {
		log.Printf("ERROR 7z %s -> %s\n", err, r.URL.RequestURI())
		e.Error = "Error unziping " + upload
		lib.SendErrorToClient(w, e)
		lib.GenericCommand(destroyTemporalDir)
		return
	}

	repoPath := dest + "/src"
	info, err := countLines(repoPath)
	if err != nil {
		log.Printf("ERROR counting loc %s -> %s\n", err, r.URL.RequestURI())
		e.Error = "Error counting LOC in " + upload
		lib.SendErrorToClient(w, e)
		lib.GenericCommand(destroyTemporalDir)
		return
	}

	resultPath := "./" + folder + "/" + folderNum + ".txt"
	lib.WriteFile(resultPath, string(info))
	readFileLineByLine(resultPath)
	//fmt.Println(languages)
	lib.SendJSONToClient(w, languages)
	lib.GenericCommand(destroyTemporalDir)
}

func doLocRepoRequest(w http.ResponseWriter, r *http.Request, repo, folder string) {
	//repouser := strings.Split(repo, "/")[0]
	folderNum := folder
	folder = "temp/loc/" + folder
	//reponame := strings.Split(repo, "/")[1]

	if !existRepo(repo) {
		e.Error = repo + " doesn't exist"
		lib.SendErrorToClient(w, e)
		return
	}
	destroyTemporalDir := []string{"rm", "-r", folder}

	createTemporalDir := []string{"mkdir", folder}
	err := lib.GenericCommand(createTemporalDir)
	if err != nil {
		log.Printf("ERROR cant create temporal dir %s\n", err)
		e.Error = "Cant create temporal dir for " + repo
		lib.SendErrorToClient(w, e)
		return
	}

	url := "https://github.com/" + repo
	dest := "./" + folder
	cloneRepo := []string{"git", "clone", url, dest}
	err = lib.GenericCommand(cloneRepo)
	if err != nil {
		log.Printf("ERROR Cant clone repo %s -> %s\n", err, r.URL.RequestURI())
		e.Error = "Can't clone repo " + repo
		lib.SendErrorToClient(w, e)
		lib.GenericCommand(destroyTemporalDir)
		return
	}
	repoPath := "./" + folder
	info, err := countLines(repoPath)
	if err != nil {
		log.Printf("ERROR counting loc %s -> %s\n", err, r.URL.RequestURI())
		e.Error = "Error counting LOC in " + repo
		lib.SendErrorToClient(w, e)
		lib.GenericCommand(destroyTemporalDir)
		return
	}

	resultPath := "./" + folder + "/" + folderNum + ".txt"
	lib.WriteFile(resultPath, string(info))
	readFileLineByLine(resultPath)
	//fmt.Println(languages)
	lib.SendJSONToClient(w, languages)
	lib.GenericCommand(destroyTemporalDir)
}

func doLocRepoRequest2(w http.ResponseWriter, r *http.Request, repo string, folder string) {
	//repouser := strings.Split(repo, "/")[0]
	folderNum := folder
	folder = "temp/loc/" + folder
	reponame := strings.Split(repo, "/")[1]

	if !existRepo(repo) {
		e.Error = repo + " doesn't exist"
		lib.SendErrorToClient(w, e)
		return
	}
	destroyTemporalDir := []string{"rm", "-r", folder}

	createTemporalDir := []string{"mkdir", folder}
	err := lib.GenericCommand(createTemporalDir)
	if err != nil {
		log.Printf("ERROR cant create temporal dir %s\n", err)
		e.Error = "Cant create temporal dir for " + repo
		lib.SendErrorToClient(w, e)
		return
	}

	url := "https://github.com/" + repo + "/archive/master.zip"
	filePath := "./" + folder + "/" + reponame + ".zip"
	err = downloadFile(filePath, url)
	if err != nil {
		log.Printf("ERROR Cant download repo %s -> %s\n", err, r.URL.RequestURI())
		e.Error = "Can't download repo " + repo
		lib.SendErrorToClient(w, e)
		lib.GenericCommand(destroyTemporalDir)
		return
	}

	dest := "./" + folder
	//unzipFile := []string{"unzip", filePath, "-d", dest + "/src"}
	unzipFile := []string{"7z", "x", filePath, "-o" + dest + "/src"}
	err = lib.GenericCommand(unzipFile)
	if err != nil {
		// some repos like rethinkdb/rethinkdb uses -next instead of -master
		url = "https://github.com/" + repo + "/archive/next.zip"
		err = downloadFile(filePath, url)
		if err != nil {
			log.Printf("Cant download repo %s\n", err)
			e.Error = "Can't download repo " + repo
			lib.SendErrorToClient(w, e)
			lib.GenericCommand(destroyTemporalDir)
			return
		}
		unzipFile = []string{"7z", "x", filePath, "-o" + dest + "/src"}
		err = lib.GenericCommand(unzipFile)
		if err != nil {
			log.Printf("ERROR 7z %s -> %s\n", err, r.URL.RequestURI())
			e.Error = "Error unziping " + repo
			lib.SendErrorToClient(w, e)
			lib.GenericCommand(destroyTemporalDir)
			return
		}
	}

	repoPath := dest + "/src"
	info, err := countLines(repoPath)
	if err != nil {
		log.Printf("ERROR counting loc %s -> %s\n", err, r.URL.RequestURI())
		e.Error = "Error counting LOC in " + repo
		lib.SendErrorToClient(w, e)
		lib.GenericCommand(destroyTemporalDir)
		return
	}

	resultPath := "./" + folder + "/" + folderNum + ".txt"
	lib.WriteFile(resultPath, string(info))
	readFileLineByLine(resultPath)
	//fmt.Println(languages)
	lib.SendJSONToClient(w, languages)
	lib.GenericCommand(destroyTemporalDir)
}

func readFileLineByLine(filePath string) {
	file, err := os.Open(filePath)
	if err != nil {
		log.Printf("ERROR opening file line by line %s\n", err)
		return
	}
	defer file.Close()
	scanner := bufio.NewScanner(file)
	number := 1
	for scanner.Scan() {
		line := scanner.Text() // string
		if number > 3 && string(line[0]) != "-" {
			processLine(line)
		}
		number++
	}
	if err := scanner.Err(); err != nil {
		log.Printf("ERROR scan file %s\n", err)
	}
}

func processLine(line string) {
	parts := strings.Split(strings.Join(strings.Fields(line), " "), " ")
	var lang language
	var index = 0
	if len(parts) > 6 { // two word language name
		lang.Name = parts[0] + " " + parts[1]
		index++
	} else {
		lang.Name = parts[0]
	}
	lang.Files = stringToInt(parts[index+1])
	lang.Lines = stringToInt(parts[index+2])
	lang.Blanks = stringToInt(parts[index+3])
	lang.Comments = stringToInt(parts[index+4])
	lang.Code = stringToInt(parts[index+5])
	languages = append(languages, lang)
}

func stringToInt(s string) int {
	num, err := strconv.Atoi(s)
	if err != nil {
		log.Printf("ERROR string to Int %s\n", err)
		return 0
	}
	return num
}

func existRepo(repo string) bool {
	url := "https://github.com/" + repo
	resp, err := http.Get(url)
	if err != nil {
		log.Printf("ERROR exists repo %s\n", err)
		return false
	}
	if resp.StatusCode == 200 {
		return true
	}
	return false
}

func downloadFile(filePath string, url string) (err error) {
	client := http.Client{
		Timeout: 30 * time.Second,
	}
	// Create the file
	out, err := os.Create(filePath)
	if err != nil {
		return err
	}
	defer out.Close()
	// Get the data
	resp, err := client.Get(url)
	if err != nil {
		return err
	}
	defer resp.Body.Close()
	// Writer the body to file
	_, err = io.Copy(out, resp.Body)
	if err != nil {
		return err
	}
	return nil
}

func countLines(repoPath string) (info []byte, err error) {
	languages = make([]language, 0)
	var e = myError{}
	info, err = exec.Command(c.Loc.LocLinux, repoPath).CombinedOutput()
	//info, err = exec.Command(c.Loc.LocMac, repoPath).CombinedOutput()
	if err != nil {
		log.Println(fmt.Sprintf("ERROR in countLines %s\n", err))
		e.Error = err.Error()
		return nil, err
	}
	return info, err
}

File: loc.js
/* */

const express = require('express');
const app = express();

const path = require('path');
const fs = require("fs");
const readline = require("readline");
const formidable = require('formidable');

const c = require(path.join(__dirname, '_config.js'));
const lib = require(path.join(__dirname, '/_lib/lib.js'));

let tokens = require(path.join(__dirname, '_private.js'));
tokens = tokens.tokens;

app.use(function (req, res, next) {
  next();
});

// localhost:3000/v1/loc?github=jolav/codetabs => {github: 'jolav/codetabs'}
app.get('/*', function (req, res) {
  //console.log('GET ==>', req.query.github);
  c.loc.counter++;
  const folder = `tmp/loc/${process.env.pm_id}-${c.loc.counter}`;
  if (req.query.github) {
    const g = req.query.github.split("/");
    if (!g[0] || g[0] === undefined || !g[1] || g[1] === undefined) {
      const msg = `user or repo is empty`;
      lib.sendError(res, msg, 400);
      return;
    }
    //repoLoc(req, res, folder);
    repoLocClone(req, res, folder);
  } else {
    const msg = `user/repo is empty`;
    lib.sendError(res, msg, 400);
  }
});

app.post('/*', function (req, res) {
  //console.log('UPLOAD');
  c.loc.counter++;
  const folder = `tmp/loc/${process.env.pm_id}-${c.loc.counter}`;
  uploadLoc(req, res, folder);
});

app.get('/*', function (req, res) {
  c.error.Error = "Bad request : " + req.path;
  res.status(400).json(c.error);
});

async function repoLoc(req, res, folder) {
  const url = "https://github.com/";
  const repo = req.query.github;
  const reponame = repo.split('/')[1];
  const destroyTemporalFolder = `rm -r ${folder}`;
  const createTemporalFolder = `mkdir ${folder}`;
  const zipFilePath = __dirname + "/" + folder + "/" + reponame + ".zip";
  const dest = __dirname + "/" + folder;
  const unzip = `7z x ${zipFilePath} -o${dest}/src`;
  const countloc = `${c.loc.countLoc} ${dest}/src`;

  const exists = await existsRepo(url + repo);
  if (!exists) {
    //console.log('NO EXISTS');
    const msg = `repo ${repo} doesn't exist`;
    lib.sendError(res, msg, 400);
    doCommand(destroyTemporalFolder);
    return;
  }
  try {
    await doCommand(destroyTemporalFolder);
    await doCommand(createTemporalFolder);
  } catch (err) {
    console.error('ERROR destroying or creating new folder ', folder, err);
    return;
  }

  // some repos like rethinkdb/rethinkdb uses -next instead of -master
  const downloadA = "https://codeload.github.com/" + repo + "/zip/master";
  const downloadB = "https://codeload.github.com/" + repo + "/zip/next";
  let isDownloaded = false;
  const existsDownload = await existsRepo(downloadA);
  if (existsDownload) {
    isDownloaded = await doDownload(downloadA, zipFilePath);
  } else {
    isDownloaded = await doDownload(downloadB, zipFilePath);
  }

  if (!isDownloaded) {
    const msg = "Can't download repo " + repo;
    lib.sendError(res, msg, 400);
    doCommand(destroyTemporalFolder);
    return;
  }

  const unzipError = await doCommand(unzip);
  if (unzipError !== null) {
    const msg = "Error unziping " + repo;
    lib.sendError(res, msg, 400);
    doCommand(destroyTemporalFolder);
    return;
  }

  const txtlocs = await doCommand(countloc);
  if (!txtlocs) {
    const msg = "Error counting LOC in " + repo;
    lib.sendError(res, msg, 400);
    doCommand(destroyTemporalFolder);
    return;
  }

  const errorTxt = await createTxt(dest + "/" + c.loc.order + ".txt", txtlocs);
  if (errorTxt !== null) {
    const msg = "Error saving data txt " + repo;
    lib.sendError(res, msg, 400);
    doCommand(destroyTemporalFolder);
    return;
  }

  //const locs = await processTxt(__dirname + "/2.txt");
  const locs = await processTxt(dest + "/" + c.loc.order + ".txt");
  if (!locs) {
    const msg = "Error counting LOC txt in " + repo;
    lib.sendError(res, msg, 400);
    doCommand(destroyTemporalFolder);
    return;
  }

  //console.log('********** END **********');
  lib.sendResult(req, res, locs, 200);
  doCommand(destroyTemporalFolder);
}

// git clone --depth=1 https://github.com/user/repo
async function repoLocClone(req, res, folder) {
  //console.log('clone repo');
  const url = "https://github.com/";
  const repo = req.query.github;
  const destroyTemporalFolder = `rm -r ${folder}`;
  const createTemporalFolder = `mkdir ${folder}`;
  const dest = __dirname + "/" + folder;
  const countloc = `${c.loc.countLoc} ${dest}/`;

  const exists = await existsRepo(url + repo);
  if (!exists) {
    //console.log('NO EXISTS');
    const msg = `repo ${repo} doesn't exist`;
    lib.sendError(res, msg, 400);
    doCommand(destroyTemporalFolder);
    return;
  }

  // limit repo size to XMB
  let size = 0;
  try {
    const options = {
      timeout: 3000,
      host: 'api.github.com',
      port: 443,
      path: '/repos/' + repo,
      method: 'GET',
      headers: {
        "User-Agent": "jolav/codetabs",
        'Accept': 'application/json',
        'Authorization': 'Bearer ' + tokens[lib.getRandomNumber(0, 9)],
        'Accept-Charset': 'utf-8'
      }
    };
    const repoData = await lib.makeRequestAsync(options, undefined);
    size = Math.floor(repoData.size / 1000);
  } catch (err) {
    console.error(err);
    const msg = `Internal error`;
    lib.sendError(res, msg, 503);
    doCommand(destroyTemporalFolder);
    return;
  }
  if (size > c.loc.maxSize) {
    const msg = `repo ${repo} too big (>${c.loc.maxSize}MB) = ${size} MB`;
    lib.sendError(res, msg, 400);
    doCommand(destroyTemporalFolder);
    return;
  }

  try {
    await doCommand(destroyTemporalFolder);
    await doCommand(createTemporalFolder);
  } catch (err) {
    console.error('ERROR destroying or creating new folder ', folder, err);
    return;
  }

  // some repos like rethinkdb/rethinkdb uses -next instead of -master
  const cloneRepoURL = "https://github.com/" + repo;
  const doClone = `git clone --depth=1 ${cloneRepoURL} ${dest}/`;
  //console.log(doClone);
  try {
    await doCommand(doClone);
  } catch (err) {
    console.error("ERROR git clone =>", err);
    const msg = "Can't clone repo " + repo;
    lib.sendError(res, msg, 400);
    doCommand(destroyTemporalFolder);
    return;
  }

  const txtlocs = await doCommand(countloc);
  if (!txtlocs) {
    const msg = "Error counting LOC in " + repo;
    lib.sendError(res, msg, 400);
    doCommand(destroyTemporalFolder);
    return;
  }

  const errorTxt = await createTxt(dest + "/" + c.loc.order + ".txt", txtlocs);
  if (errorTxt !== null) {
    const msg = "Error saving data txt " + repo;
    lib.sendError(res, msg, 400);
    doCommand(destroyTemporalFolder);
    return;
  }

  //const locs = await processTxt(__dirname + "/2.txt");
  const locs = await processTxt(dest + "/" + c.loc.order + ".txt");
  if (!locs) {
    const msg = "Error counting LOC txt in " + repo;
    lib.sendError(res, msg, 400);
    doCommand(destroyTemporalFolder);
    return;
  }

  //console.log('********** END **********');
  lib.sendResult(req, res, locs, 200);
  doCommand(destroyTemporalFolder);
}

async function uploadLoc(req, res, folder) {
  //console.log('UPLOAD',c.loc.orderInt);
  const destroyTemporalFolder = `rm -r ${folder}`;
  const createTemporalFolder = `mkdir ${folder}`;
  const dest = __dirname + "/" + folder;
  const countloc = `${c.loc.countLoc} ${dest}/src`;

  try {
    await doCommand(destroyTemporalFolder);
    await doCommand(createTemporalFolder);
  } catch (err) {
    console.error('ERROR destroying or creating new folder ', folder, err);
  }

  const fileUploaded = await uploadFile(req, folder);
  const zipFilePath = folder + "/" + fileUploaded;
  if (fileUploaded === null) {
    const msg = "Error uploading file";
    lib.sendError(res, msg, 400);
    doCommand(destroyTemporalFolder);
    return;
  }

  //unzipFile := []string{"7z", "x", filePath, "-o" + dest + "/src"}
  const unzip = `7z x ${zipFilePath} -o${dest}/src`;
  const unzipError = await doCommand(unzip);
  if (unzipError !== null) {
    const msg = "Error unziping " + fileUploaded;
    lib.sendError(res, msg, 400);
    doCommand(destroyTemporalFolder);
    return;
  }

  const txtlocs = await doCommand(countloc);
  if (!txtlocs) {
    const msg = "Error counting LOC in " + fileUploaded;
    lib.sendError(res, msg, 400);
    doCommand(destroyTemporalFolder);
    return;
  }

  const errorTxt = await createTxt(dest + "/" + c.loc.orderInt + ".txt", txtlocs);
  if (errorTxt !== null) {
    const msg = "Error saving data txt " + fileUploaded;
    lib.sendError(res, msg, 400);
    doCommand(destroyTemporalFolder);
    return;
  }

  //const locs = await processTxt(__dirname + "/2.txt");
  const locs = await processTxt(dest + "/" + c.loc.orderInt + ".txt");
  if (!locs) {
    const msg = "Error counting LOC txt in " + fileUploaded;
    lib.sendError(res, msg, 400);
    doCommand(destroyTemporalFolder);
    return;
  }
  //console.log('********** END **********');
  lib.sendResult(req, res, locs, 200);
  doCommand(destroyTemporalFolder);
}

function uploadFile(req, folder) {
  return new Promise((resolve, reject) => {
    const form = new formidable.IncomingForm();
    form.parse(req, function (err, fields, files) {
      const filename = files.inputFile.name;
      const oldpath = files.inputFile.path;
      const newpath = folder + "/" + filename;
      fs.rename(oldpath, newpath, function (err) {
        if (err) {
          console.log('Error => ', err);
          resolve();
        }
        resolve(filename);
      });
    });
  });
}

function processTxt(textFile) {
  return new Promise((resolve, reject) => {
    countLocFromTxt(textFile, function (err, locs) {
      if (err) {
        resolve(err);
      }
      resolve(locs);
    });
  });
}

function countLocFromTxt(textFile, cb) {
  const lineReader = readline.createInterface({
    input: fs.createReadStream(textFile)
  });
  const languages = [];
  let lang = {
    'language': "language",
    'files': "files",
    'lines': "lines",
    'blanks': "blanks",
    'comments': "comments",
    'linesOfCode': "linesOfCode"
  };
  let lineNumber = 1;
  lineReader.on('line', function (line) {
    if (lineNumber > 3 && line[0] !== "-") {
      let parts = line.trim().replace(/\s\s+/g, ' ').split(" ");
      //console.log(parts);
      let index = 0;
      lang = {};
      if (parts.length > 6) { // two word language name 
        lang.language = parts[0] + " " + parts[1];
        index++;
      } else {
        lang.language = parts[0];
      }
      lang.files = parts[index + 1];
      lang.lines = parts[index + 2];
      lang.blanks = parts[index + 3];
      lang.comments = parts[index + 4];
      lang.linesOfCode = parts[index + 5];
      //console.log('lang',lang);
      languages.push(lang);
    }
    lineNumber++;
  });
  lineReader.on('close', function () {
    cb(languages);
  });
}

function createTxt(filePath, data) {
  return new Promise((resolve, reject) => {
    fs.writeFile(filePath, data, function (err) {
      if (err) {
        resolve(err);
      }
      resolve(null);
    });
  });
}

function doDownload(url, zipFilePath) {
  return new Promise((resolve, reject) => {
    lib.downloadFile(url, zipFilePath, function (isDone) {
      resolve(isDone);
    });
  });
}

function doCommand(command) {
  //console.log('Command ...', command);
  return new Promise((resolve, reject) => {
    lib.linuxCommand(command, function (err, res) {
      if (err) {
        //console.log('Error doCommand => ', command);
        //console.log(err);
        //reject(res);
      }
      if (command.split(" ")[0] === c.loc.countLoc) {
        resolve(res);
      } else {
        resolve(err);
      }
    });
  });
}

function existsRepo(urlchecked) {
  return new Promise((resolve, reject) => {
    lib.checkUrlExists(urlchecked, function (res) {
      resolve(res);
    });
  });
}

module.exports = app;
